tools/wds2idx.py

```diff
@@ -0,0 +1,185 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import time
+import argparse
+import subprocess
+
+
+class IndexCreator:
+    """Reads `Webdataset` data format, and creates index file
+    that enables random access.
+
+    Example usage:
+    ----------
+    >>> creator = IndexCreator('data/test.tar','data/test.idx')
+    >>> creator.create_index()
+    >>> creator.close()
+    >>> !ls data/
+    test.tar  test.idx
+
+    Parameters
+    ----------
+    uri : str
+        Path to the archive file.
+    idx_path : str
+        Path to the index file, that will be created/overwritten.
+    """
+
+    def __init__(self, uri, idx_path):
+        self.uri = uri
+        self.idx_path = idx_path
+        self.fidx = open(self.idx_path, "w")
+
+    def open(self):
+        """Opens the archive and index files and sets their read heads to 0."""
+        if self.fidx.closed:
+            self.fidx = open(self.idx_path, "w")
+        else:
+            self.fidx.seek(0)
+
+    def close(self):
+        """Closes the archive and index files."""
+        if not self.fidx.closed:
+            self.fidx.close()
+
+    def reset(self):
+        """Resets the archive and index files."""
+        self.close()
+        self.open()
+
+    @staticmethod
+    def split_name(filepath):  # translated from the matching function in c++
+        """Splits the webdataset into the basename and the extension"""
+        base_name_pos = filepath.rfind("\\") + 1
+        dot_pos = filepath.find(".", base_name_pos + 1)
+        if dot_pos == -1:
+            return filepath, ""
+        return filepath[:dot_pos], filepath[dot_pos + 1 :]
+
+    def create_index(self):
+        """Creates the index file from open record file"""
+        self.reset()
+
+        pre_time = time.time()
+        counter = 0
+        report_step = 100000
+
+        print(f"time: {time.time() - pre_time:.2f} count: {counter} stage: collect")
+
+        # Collects the data about the contents of the tar archive
+        tar_blocks_proc = subprocess.Popen(
+            ["tar", "--list", "--block-num", "--file", self.uri],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        tar_types_sizes_proc = subprocess.Popen(
+            ["tar", "--verbose", "--list", "--file", self.uri],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
```

Comment:
What if the user doesn't have tar available? Maybe we should fallback to the python tart in that case?

Reply:
hmmm, alright

Reply:
done